scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamWriter.scala

```diff
@@ -4,14 +4,21 @@ import com.stripe.dagon.Rule
 import com.twitter.scalding.Execution.{ToWrite, Writer}
 import com.twitter.scalding.typed._
 import com.twitter.scalding.{CFuture, CancellationHandler, Config, Execution, ExecutionCounters}
+import java.nio.channels.Channels
 import java.util.concurrent.atomic.AtomicLong
 import org.apache.beam.sdk.Pipeline
+import org.apache.beam.sdk.coders.Coder
+import org.apache.beam.sdk.io.FileSystems
 import scala.annotation.tailrec
 import scala.concurrent.{ExecutionContext, Future}
 
 class BeamWriter(val beamMode: BeamMode) extends Writer {
   private val state = new AtomicLong()
 
+  private val sourceCounter: AtomicLong = new AtomicLong(0L)
+  private val iterablePipes: scala.collection.mutable.Map[TypedPipe[_], (Coder[_], String)] =
```

Comment:
I think this needs to be thread-safe, so it may be best to use `ConcurrentHashMap` from java or use a mutex somehow.

Reply:
Done